[SPARK-49566][SQL] Add SQL pipe syntax for the EXTEND operator

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for the EXTEND operator.

This operator preserves the existing input table and adds one or more new computed columns  whose values are equal to evaluating the specified expressions. This is equivalent to `SELECT *, <newExpressions>` in the SQL compiler. It is provided as a convenience feature and some functionality overlap exists with lateral column aliases.

For example:

```
CREATE TABLE t(x INT, y STRING) USING CSV;
INSERT INTO t VALUES (0, 'abc'), (1, 'def');

TABLE t
|> EXTEND x + LENGTH(y) AS z;

+----+-----+-----+
|  x |   y |   z |
+----+-----+-----+
|  0 | abc |   3 |
|  1 | def |   4 |
+----+-----+-----+
```

Like the `|> SELECT` operator, aggregate functions are not allowed in these expressions. During the course of developing reasonable error messages for this, I found that the SQL pipe syntax research paper also specified that the `|> AGGREGATE` operator should require that each non-grouping expression contains at least one aggregate function; I added a check and reasonable error message for this case as well.

### Why are the changes needed?

The SQL pipe operator syntax will let users compose queries in a more flexible fashion.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds a few unit test cases, but mostly relies on golden file test coverage. I did this to make sure the answers are correct as this feature is implemented and also so we can look at the analyzer output plans to ensure they look right as well.

### Was this patch authored or co-authored using generative AI tooling?

No

dtenedor committed:
commit

commit

commit

commit

commit

commit

commit

commit

fix tests

commit

commit

commit

commit

commit

commit

dtenedor committed:
Merge branch 'master' into pipe-syntax-projections
sync

dtenedor committed:
fix test

dtenedor committed:
split off set, drop, as operator from this PR for now

dtenedor committed:
split off set, drop, as operator from this PR for now

dtenedor commented:
This PR now just covers the `EXTEND` operator.

dtenedor committed:
split off set, drop, as operator from this PR for now

dtenedor commented:
cc @gengliangwang @cloud-fan this one is ready for review at your convenience.

dtenedor committed:
split off set, drop, as operator from this PR for now

dtenedor committed:
split off set, drop, as operator from this PR for now

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/pipeOperators.scala:
```diff
@@ -18,26 +18,26 @@
+ * |> EXTEND. We use this to make sure that no aggregate functions exist in these expressions.
  */
-case class PipeSelect(child: Expression)
+case class PipeSelect(child: Expression, clause: String = "SELECT")
```
now the name `PipeSelect` becomes a bit weird. How about just `case class PipeExpression(..., isAgg: Boolean, clause: String)`

cloud-fan replied:
`isAgg = true` means the expression must contain aggregate function, false means must not contain agg func

dtenedor replied:
Sure, sounds good, this is done.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/pipeOperators.scala:
```diff
@@ -18,26 +18,45 @@
+    PipeExpression(newChild, isAggregate, clause)
+  override lazy val replacement: Expression = if (isAggregate) {
+    // Make sure the child expression contains at least one aggregate function.
+    var foundAggregate = false
```
nit: `val hasAggFunc = this.exists(_.isInstanceOf[AggregateFunction])`

dtenedor committed:
Merge branch 'master' into pipe-syntax-projections
sync

dtenedor committed:
fix test

dtenedor reviewed (commented):
Thanks @cloud-fan  for your reviews!!

cloud-fan commented:
thanks, merging to master!
